Optimize api calls between envoy_reader and Home Assistant

gtdiehl opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The current sensor code makes one API call for each inverter installed in the Enphase solar system. For Enphase systems the solar panel and inverter have a 1-to-1 relationship. Most systems have at least 8 solar panels.  I have seen some systems with 32 and in extreme cases 300+. These multiple calls for inverter data is actually repetitive because the page containing the data is all or nothing. As well some older Envoy devices can become unresponsive and not return data when too many requests are sent to it.

So this change on the Home Assistant sensor side is to make one http call for all of the inverters and display the individual inverter data in Home Assistant using the coordinator

[envoy_reader changelog@0.17.3...0.18.2](https://github.com/jesserizzo/envoy_reader/compare/0.17.3...0.18.2)
Changes to the `envoy_reader` library include:
- Remove `update()` method that Home Assistant previously used
- Change to raising exceptions when errors occur
- Replaced `requests_async` library with `httpx`
- Move all I/O operations to a single function `getData()`
- Automatic retry when gathering Inverter data. Moving to `httpx` seems to uncovered an underlying issue with older Envoy devices

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
sensor:
  - platform: enphase_envoy
    ip_address: !secret enphase_ip_addr
    username: !secret enphase_user
    password: !secret enphase_pass
    monitored_conditions:
      - production
      - daily_production
      - seven_days_production
      - lifetime_production
      - inverters
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fix #30425 jesserizzo/envoy_reader#12
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

gtdiehl committed:
Updating sensor to use single API call

gtdiehl committed:
Updated comment

Updated comment to reflect that polling is needed.

gtdiehl committed:
Reduced calls to single API call

gtdiehl committed:
Added except handling and increased async timeout

gtdiehl committed:
Cleaned up some comments

gtdiehl committed:
Added error handling

gtdiehl committed:
Added last_reported date for inverters

gtdiehl committed:
Added message during failed update

gtdiehl committed:
Added retries to update function

gtdiehl committed:
Updated update function

gtdiehl committed:
Reformatted sensor.py with black

gtdiehl committed:
Increased default scan period

gtdiehl committed:
fixed timedelta typo

exxamalte commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -141,28 +207,62 @@ def device_state_attributes(self):
             else:
-                self._state = None
+                _LOGGER.debug(
+                    "Data inverter (%s) isInstance(dict) was %s.  Using previous state: %s",
```
Can I suggest a slightly less technical message, such as "Data from inverter %s received in unsupported format: %s. Using ..."

gtdiehl replied:
I figured for a debug message it didn't really need to be too readable.  Updated the message anyway 😄

exxamalte commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -141,28 +207,62 @@ def device_state_attributes(self):
+                    self._state,
+                )
+        else:
+            _LOGGER.error("Update loop - API returned invalid data!")
```
I can't see how you would ever reach this statement.

gtdiehl replied:
Removed `else` condition

exxamalte commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -141,28 +207,62 @@ def device_state_attributes(self):
-                _LOGGER.error(_state)
-                self._state = None
+                _LOGGER.debug(
+                    "Sensor %s isInstance(int) was %s.  Returning None for state.",
```
Can I suggest a slightly less technical message, such as "Data for sensor %s is not a number. ..."

gtdiehl replied:
I figured for a debug message it didn't really need to be too readable.  Updated the message anyway 😄

exxamalte reviewed (commented):
Thanks for the optimisations. Please have a look at my comments.

gtdiehl committed:
importing CoordinatorEntity

gtdiehl commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -141,28 +207,62 @@ def device_state_attributes(self):
+            self.coordinator.async_add_listener(self.async_write_ha_state)
+        )
+
+    @Throttle(MIN_SCAN_INTERVAL)
```
Removed @Throttle and the import as I am inheriting `CoordinatorEntity`

gtdiehl commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -86,33 +131,54 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
 
-    def __init__(self, envoy_reader, sensor_type, name, unit):
-        """Initialize the sensor."""
+    def __init__(self, envoy_reader, sensor_type, name, unit, coordinator):
```
Removed access to `envoy_reader` as it is not needed

gtdiehl committed:
Check during setup else raise PlatformNotReady

gtdiehl commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -51,33 +58,71 @@
-                )
-                continue
-
+            inverters = coordinator.data.get("inverters_production")
```
Added a check if the data is `NoneType` than `raise PlatformNotReady`

gtdiehl replied:
- Removed `should_poll`
- Removed `available`
- Removed `async_added_to_hass`
- Removed `async_update` and moved the code into `state`
- In `state` it wasn't a copy and paste from `async_update`, but mostly the same code exists. Tried to reduce the paths and catch the same errors. I moved the update for the inverter `last_reported` attribute into the `device_state_attributes()` method

I noticed now the sensors have values on the first poll rather than waiting for the second poll!

gtdiehl commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -86,33 +131,54 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
 
 class Envoy(Entity):
-    """Implementation of the Enphase Envoy sensors."""
+    """Envoy entity."""
```
Inherited CoordinatorEntity, is this all I have to do?

exxamalte replied:
There is a bit more:
* No need to override `should_poll`.
* No need to override `available`.
* No need to override `async_added_to_hass`.
* And you want to transform the code from `async_update` into overriding `state`. (There is a high-level introduction into `CoordinatorEntity` at https://developers.home-assistant.io/docs/integration_fetching_data/)

exxamalte replied:
Oh, and you only need to inherit from `CoordinatorEntity`, not from `Entity` anymore.

gtdiehl replied:
Sorry put the comment in the wrong section! 😮 

- Removed should_poll
- Removed available
- Removed async_added_to_hass
- Removed async_update and moved the code into state
- In state it wasn't a copy and paste from async_update, but mostly the same code exists. Tried to reduce the paths and catch the same errors. I moved the update for the inverter last_reported attribute into the device_state_attributes() method

I noticed now the sensors have values on the first poll rather than waiting for the second poll!

gtdiehl committed:
Removed async_update and override state

exxamalte reviewed (approved):
Looks good to me now.

frenck commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -51,67 +61,121 @@
             cv.ensure_list, [vol.In(list(SENSORS))]
         ),
         vol.Optional(CONF_NAME, default=""): cv.string,
+        vol.Optional(CONF_SCAN_INTERVAL, default=60): cv.time_period,
```
Scan interval is already part of the platform schema you are extending here.

To set a different scan interval on this platform, only the `SCAN_INTERVAL` constant needs to be set.

gtdiehl replied:
I set the `SCAN_INTERVAL` constant and removed the `CONF_SCAN_INTERVAL` from the `PLATFORM_SCHEMA`.  I'm just using the `SCAN_INTERVAL` constant directly when setting the `update_interval` attribute for the `coordinator`. 

I know [ADR-0010](https://github.com/home-assistant/architecture/pull/366) says that changes to YAML configurations will no longer be excepted. In the future how can I make the `SCAN_INTERVAL` configurable?

frenck replied:
Well, essentially, we don't want them to be configurable. It should have a sane default.
Nevertheless, if there is a good reason to still make them configurable, if migrated to a configuration flow (ADR-0010), one can add an OptionFlow to allow for setting that.

gtdiehl replied:
All my changes are made.  Thank you for your help!

gtdiehl committed:
using SCAN_INTERVAL constant

gtdiehl committed:
fixed typo

gtdiehl committed:
removed unused constant

bdraco commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -65,53 +76,103 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
+            i = 0
+            while i < 3:
+                try:
+                    data = await envoy_reader.update()
```
Can this logic be moved into the upstream library?

gtdiehl replied:
I don't think I even need to have it in the envoy_reader library. The sensor retrieves data every 60seconds. Envoy hardware devices update anywhere from real-time to up-to every 15mins, depending on the model. If one polling cycle is missed due to some network communication problem or maybe the device can't handle the request (which I hope won't happen as often anymore after this PR gets merged!).

I can remove the while loop and just raise UpdateFailed(...).

gtdiehl replied:
@bdraco made my code change

gtdiehl committed:
Removed retry logic

bdraco commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -65,53 +76,94 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
+
+            _LOGGER.debug("Retrieved data from API: %s", data)
+
+            if "can't handle event type ConnectionClosed" in str(
+                data.get("inverters_production")
+            ):
+                raise UpdateFailed("Inverter updated failed")
+
```
Can we use a typed exception here instead to avoid the string matching?

gtdiehl replied:
The api being used is returning a string with the exception inside that. I know not the best solution but will need to be addressed. Without changing the library. I only added these because they were not being caught in the previous sensor code. The api is returning the same as it was previously. I can remove the string checks and go back to the previous behavior of the sensor code which was to really not check much?

bdraco replied:
While it's not ideal, it's preferable for the library to do string matching and throw the typed exception if you don't have another way to handle it.

That way if the api changes we only have to bump the library and don't have to wait for a free reviewer on the HomeAssistant side

gtdiehl replied:
Let me see what I can come up with in a few days.

gtdiehl replied:
I have changed the `async_update_data()` method to use try/except blocks rather than string matching.
Take a look at the updated code. I had to change the library to raise exceptions rather than returning the strings.

The api change log is [here](https://github.com/jesserizzo/envoy_reader/compare/master...gtdiehl:exception_handling_ha?expand=1)

I will release the updated library tomorrow and open a separate PR here to bump the `enphase_envoy` library to `0.17.1`

bdraco commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -65,53 +76,94 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
+                        f"{name}{SENSORS[condition][0]} {inverter}",
+                        condition,
+                    )
+            elif "Unable to connect to Envoy" in inverters:
+                _LOGGER.error(
+                    "Unable to connect to Enphase Envoy during setup. Inverter entities not added. Please check IP address and credentials are correct."
+                )
+            elif "can't handle event type ConnectionClosed" in inverters:
+                _LOGGER.error(
+                    "Communication error with Enphase Envoy during setup. Inverter entities not added."
+                )
```
Can we use a typed exceptions here instead to avoid the string matching?

gtdiehl replied:
The api being used is returning a string with the exception inside that. I know not the best solution but will need to be addressed. Without changing the library. I only added these because they were not being caught in the previous sensor code. The api is returning the same as it was previously. I can remove the string checks and go back to the previous behavior of the sensor code which was to really not check much?

gtdiehl replied:
Changed to using exceptions rather than string matching.

bdraco reviewed (changes requested):
Please see comments above about avoiding string matching and using typed exeptions.

gtdiehl commented:
> Please see comments above about avoiding string matching and using typed exeptions.

I added my comments. I can remove the checks and revert back to previous sensor behavior?

gtdiehl committed:
Changed to catching exceptions rather than strings

gtdiehl committed:
shortened string split line

bdraco commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -1,8 +1,12 @@
 from envoy_reader.envoy_reader import EnvoyReader
-import requests
+import httpx
+from requests_async.exceptions import HTTPError, RequestException
```
`requests_async` is deprecated. We should not use it in new code.  Please use the httpx exceptions intead.

https://github.com/encode/requests-async

gtdiehl replied:
Replaced `requests_async` with `httpx`.

I also did this in the `envoy_reader` library so the version for the library has bumped to 0.17.2

Here is the [change log](https://github.com/jesserizzo/envoy_reader/compare/0.17.0...0.17.2) for the envoy_reader

gtdiehl replied:
I bumped the envoy_reader  to `0.18.2` as after I moved from `requests_async` to `httpx`, I was seeing `httpx.RemoteProtocolError` exceptions from older Envoy devices. I added a retry in the library so it should hopefully eliminate or at least minimize the exceptions occurring on older devices.

bdraco reviewed (changes requested):
Please see comment about `requests_async` above

MartinHjelmare commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -121,7 +185,37 @@ def name(self):
+
+        elif self._type == "inverters":
+            split_name = self._name.split(" ")
+            serial_number = split_name[(len(split_name) - 1)]
```
Since we need the inverter serial number in the entity, pass it as a parameter when creating the entity. Saves us doing this op here.

gtdiehl replied:
Made this change in both places it existed

MartinHjelmare commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -121,7 +185,37 @@ def name(self):
+            try:
+                value = self.coordinator.data.get(self._type)
+            except AttributeError:
+                _LOGGER.debug("Data for sensor %s is not a number.", self._type)
```
We don't allow logging, a side effect, in entity state properties. If you want to keep the logging the data lookup needs to move to the state update callback that is called when the entity is told to update by the coordinator.

https://github.com/home-assistant/core/blob/f9bc7bacb6a6fa3e1bda59c5e68dbbd76d92b392/homeassistant/helpers/update_coordinator.py#L242-L245

gtdiehl replied:
Removed logging, the only logging now is within the `async_setup_platform()` and `async_update_data()` methods

MartinHjelmare commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -137,32 +231,16 @@ def icon(self):
+                value = self.coordinator.data.get("inverters_production").get(
+                    serial_number
+                )[1]
+                return {"last_reported": value}
```
Please only wrap the line that can raise in the try... except block. Move this down to the end.

gtdiehl replied:
Removed try/except block from other comment.

MartinHjelmare commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -137,32 +231,16 @@ def icon(self):
+                _LOGGER.debug(
+                    "Data received from inverter for last reported time was an unsupported format."
                 )
+                return
```
```suggestion
                return None
```

MartinHjelmare commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -121,7 +185,37 @@ def name(self):
+                    "Data received from inverter %s was an unsupported format.",
+                    serial_number,
+                )
+                return
```
```suggestion
                return None
```

MartinHjelmare commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -121,7 +185,37 @@ def name(self):
+                value = self.coordinator.data.get(self._type)
+            except AttributeError:
+                _LOGGER.debug("Data for sensor %s is not a number.", self._type)
+                return
```
```suggestion
                return None
```

gtdiehl committed:
Replace requests_async with httpx

gtdiehl committed:
Bump envoy_reader version to 0.17.2

gtdiehl committed:
Resolving comments from PR requested changes

gtdiehl committed:
Fixed typo in scan_interval

gtdiehl committed:
Removed period from logging messages

gtdiehl commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -65,53 +77,105 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
+            and coordinator.data.get("inverters_production") is not None
+        ):
+            inverters = coordinator.data.get("inverters_production")
+            coordinator.data["inverters_production"] = inverters
```
Removed this part it was not needed. I didn't see that! 😅

gtdiehl commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -65,53 +77,105 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
 
+    async def async_update_data():
+        """Fetch data from API endpoint."""
+        data = dict()
```
Made change

gtdiehl commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -121,7 +185,37 @@ def name(self):
+        if self._type != "inverters":
+            try:
+                value = self.coordinator.data.get(self._type)
+            except AttributeError:
```
Removed the try/except blocks for the checking of the `AttributeError` in several places. Also for inverters I just set the data to `None` during update if a failure occurs.

gtdiehl committed:
Bumping envoy_reader to 0.18.0

gtdiehl committed:
Merge branch 'dev' into http_api_call

MartinHjelmare commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -65,53 +76,97 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
+
+            for condition in monitored_conditions:
+                if condition != "inverters":
+                    data[condition] = await getattr(envoy_reader, condition)()
```
Does this call not do I/O?

gtdiehl replied:
This call should not do I/O. The `getData()` method should be the only method that causes I/O.

But looking into the code `envoy_reader` code there is still a path that does a HTTP request.  I'll fix it in the library

gtdiehl replied:
Bumped the API version to 0.18.1. Fixed the I/O issue, so only getData() and no other methods called from the Home Assistant sensor cause I/O

MartinHjelmare commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -65,53 +76,97 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
-
-        else:
+                )
+                _LOGGER.debug(
```
Just make one logging call at the end and pass all entities to the log message. They have a repr magic method. Ie, don't log in the for loop.

gtdiehl replied:
I removed this debug logging as it's not really needed.

MartinHjelmare commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -121,7 +176,23 @@ def name(self):
+            and self.coordinator.data.get("inverters_production") is not None
+        ):
+            split_name = self._name.split(" ")
+            serial_number = split_name[-1]
```
Pass the serial number as a parameter to the entity.

gtdiehl replied:
I grab the serial number when the array is created that is passed to `async_add_entities`, which is than passed as a parameter to the entity

gtdiehl committed:
Incorporating suggested changes

MartinHjelmare commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -55,63 +64,112 @@
+                        data["inverters_production"] = await getattr(
+                            envoy_reader, "inverters_production"
+                        )()
+                    except httpx.HTTPStatusError as err:
```
Can we remove this error handling if there's no I/O left in this library call?

gtdiehl replied:
Yeah, removed this error handling it's not needed

gtdiehl committed:
Removing no longer used try/except

MartinHjelmare commented on homeassistant/components/enphase_envoy/sensor.py:
```diff
@@ -55,63 +64,103 @@
+            try:
+                await envoy_reader.getData()
+            except httpx.HTTPStatusError as err:
+                _LOGGER.error("Error retrieving data: %s", err)
```
Why do we not raise `UpdateFailed` in this case?

gtdiehl replied:
Inverters are the only monitored conditions that require authentication. My thought on this was to log an error only when authentication fails and do UpdateFailed for other errors. So that we continue to get data for the other conditions. I wasn't sure of how other integrations handle this. If others do an all or nothing approach I can change it from an error to UpdateFailed.

MartinHjelmare replied:
If this is the only call that does I/O and it fails there won't be any new data. Am I missing something?

gtdiehl replied:
Yeah, this is the only I/O call. The `getData()` call makes a batch retrieval from the device, then when the entity updates its' state we can still get the state of the data that was successful. I guess the side effect with the current implementation here is that every poll (60 seconds) an error is logged saying the Inverter retrieval received a 401 error.

If I go with only UpdateFailed no entities will be created upon setup and eventually will stop trying to setup up the platform after all the retries.

I guess the appropriate behavior is to just fail the setup completely and let the user either fix their `username`/`password` or remove `inverters` from `monitored_conditions` in the` configuration.yaml.`?

MartinHjelmare replied:
We should fail platform set up if there's an auth error and let the user fix the auth credentials.

gtdiehl replied:
I added this to check if the connection can be made to the device or not. If authentication fails, the setup fails and does not continue. If an HTTP error occurs then let HA retry and recover from the transient state.

https://github.com/home-assistant/core/blob/778a82959f27561fe716700f18a709d33f9a35c1/homeassistant/components/enphase_envoy/sensor.py#L82-L88

gtdiehl committed:
Fail setup if authentication fails

gtdiehl committed:
Bump envoy_reader to 0.18.2

gtdiehl commented:
I bumped the envoy_reader to `0.18.2` as after I moved from `requests_async` to `httpx` as requested by @bdraco, I was seeing `httpx.RemoteProtocolError` exceptions from some older Envoy devices. I added a retry in the library so it should hopefully eliminate or at least minimize the exceptions occurring on these older devices.

MartinHjelmare reviewed (approved):
Thanks!

MartinHjelmare commented:
Please add a link to a changelog or GitHub commit compare view for the version bump in the PR description. Thanks!

gtdiehl commented:
Added a link and list of changes to the library in the PR description. Thank you! 😃

MartinHjelmare commented:
@bdraco ok to merge?

bdraco commented:
Will test shortly

bdraco commented:
<img width="625" alt="Screen Shot 2020-12-28 at 12 42 18 PM" src="https://user-images.githubusercontent.com/663432/103247207-2d184d00-490a-11eb-9e31-68f09a483602.png">
<img width="587" alt="Screen Shot 2020-12-28 at 12 42 08 PM" src="https://user-images.githubusercontent.com/663432/103247208-2f7aa700-490a-11eb-93dd-59eb1bcebc52.png">

Appears to be working as expected.

bdraco reviewed (approved):
Lgtm

I couldn't get it to work with the old 3.x firmware but that's not a new issue.

These are discoverable via zeroconf so converting to config flow would be a nice next step.

bdraco commented:
@gtdiehl Pointed me to a link to find the installer password which allowed it to work on the 3.x firmware.  Confirmed working on both 5.x and 3.x 👍
